pkg/monitor/controller_test.go

```diff
@@ -0,0 +1,719 @@
+		namespaceInformer,
+		serviceInformer,
+		networkPolicyInformer,
+		nodeInformer,
+		cnpInformer,
+		anpInformer,
+		tierInformer,
+		cgInformer,
+		grpInformer,
+		addressGroupStore,
+		appliedToGroupStore,
+		networkPolicyStore,
+		groupStore,
+		false)
+
+	controllerQuerier := querier.NewControllerQuerier(networkPolicyController, 10349)
+	externalNodeEnabled := true
+	controllerMonitor := NewControllerMonitor(crdClient, nodeInformer, externalNodeInformer, controllerQuerier, externalNodeEnabled)
+
+	return &fakeController{
+		controllerMonitor,
+		crdClient,
+		client,
+		informerFactory,
+		crdInformerFactory,
+	}
+}
+
+func initController(controller *fakeController, stopCh chan struct{}) {
+	controller.informerFactory.Start(stopCh)
+	controller.crdInformerFactory.Start(stopCh)
+	controller.informerFactory.WaitForCacheSync(stopCh)
+	controller.crdInformerFactory.WaitForCacheSync(stopCh)
+}
+
+func TestSyncExternalNode(t *testing.T) {
+	en := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "TestExternalNode",
+			Namespace: "ns1",
+			Labels:    map[string]string{"en": "vm2"},
+		},
+	}
+	tc := []struct {
+		name                 string
+		key                  string
+		existingExternalNode *v1alpha1.ExternalNode
+		expectedAgentCRD     *v1beta1.AntreaAgentInfo
+		expectedError        string
+	}{
+		{
+			name:          "Invalid key format",
+			key:           "namespace/name/error",
+			expectedError: "unexpected key format: \"namespace/name/error\"",
+		},
+		{
+			name:          "Key does not exist",
+			key:           "default/vm2-e8be5",
+			expectedError: "",
+		},
+		{
+			name:                 "key exists",
+			key:                  "ns1/TestExternalNode",
+			existingExternalNode: en,
+			expectedAgentCRD: &v1beta1.AntreaAgentInfo{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            "TestExternalNode",
+					ResourceVersion: "",
+					Generation:      0,
+				},
+				NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+					NetworkPolicyNum:  0,
+					AddressGroupNum:   0,
+					AppliedToGroupNum: 0,
+				},
+			},
+			expectedError: "",
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			clientset := fakeclientset.NewSimpleClientset()
+			if tt.existingExternalNode != nil {
+				clientset = fakeclientset.NewSimpleClientset(tt.existingExternalNode)
+			}
+			controller := newControllerMonitor(clientset)
+			stopCh := make(chan struct{})
+			initController(controller, stopCh)
+			defer close(stopCh)
+			err := controller.controllerMonitor.syncExternalNode(tt.key)
+			if tt.expectedError != "" {
+				assert.Equal(t, tt.expectedError, err.Error())
+			} else {
+				require.NoError(t, err)
+				if tt.expectedAgentCRD != nil {
+					crd, err := controller.controllerMonitor.client.CrdV1beta1().AntreaAgentInfos().Get(ctx, tt.expectedAgentCRD.Name, metav1.GetOptions{})
+					require.NoError(t, err)
+					assert.Equal(t, tt.expectedAgentCRD, crd)
+				}
```

Comment:
in the case where `tt.expectedAgentCRD = nil`, maybe you could still call `Get` and check that you get a NotFound error? That would apply for the `"Key does not exist"` test case.